sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/typeCheckHelpers.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.sql.types._
+
+trait NumericHolder {
+  self: Expression =>
+
+  protected def baseType: DataType = dataType
+
+  lazy val numeric = baseType match {
+    case n: NumericType => n.numeric.asInstanceOf[Numeric[Any]]
+    case NullType => UnresolvedNumeric
+  }
+}
+
+trait OrderingHolder {
+  self: Expression =>
+
+  protected def baseType: DataType = dataType
+
+  lazy val ordering = baseType match {
+    case a: AtomicType => a.ordering.asInstanceOf[Ordering[Any]]
+    case NullType => UnresolvedOrdering
+  }
+}
+
+object NumericTypeChecker extends (DataType => Boolean) {
```

Comment:
instead of this, I'd just create a TypeUtils class in sql.catalyst.util package, and then define 3 functions
```
object TypeUtils {
  def validForBitwiseExpr(..): Boolean

  def isNumeric(...): Boolean

  def validForOrdering(..): Boolean
}
```